### PR TITLE
CDDA merge (0/x): overall review and merge plan

### PR DESCRIPTION
**Pull request by krcroft**
... add support for MP3, FLAC, and Opus; add handling of mono and non-44.1KHz tracks; and improve internal handling of CD audio streams.

**Notable changes (perhaps worthy of separate reviews):**
- added build script
- grouped single CI build into operating system workflows
- used build script in CI builds for wider compiler and release type coverage
- improved DOSBox's internal CDROM image audio handling
- replaced dependence on SDL_Sound with internal handling
- replaced dependence on libvorbis with built-in `stb_vorbis` codec
- replaced dependence on libFLAC (which never worked) with built-in `dr_flac` codec
- replaced dependence on libMPEG123 (which never worked) with build-in `dr_mp3` codec
    - added PCM-exact seeking through the use of a cached lookup table
- replaced dependence on SDL_Sound WAV codec with more robust built-in `dr_wav` codec
- added Opus handling via pkg-config instead of internal handling via Makefile
    - allows build-compatibility with VisualStudio (with all codecs enabled)

Changes have been build-and-play tested on the following operating systems: Windows 10, Ubuntu Linux 18.04 and 19.04, Rasbian; as well as x86_64, i686, ARM Cortex-A53, PowerPC 7400 (Big-Endian) architectures.

### CODE REVIEW

**Comment by krcroft** [comment]
Says it's waiting for two CI jobs, but they've all passed. Seems like a bug?

**Comment by dreamer** [comment]
This commit affected build names, and those 2 builds were marked as "required to pass" in branch configuration - I unchecked them for now.

In normal circumstances I like kernel and git project style of review/accepting patches, that forces split into ~one commit per function, but these are not normal circumstances - and we don't have manpower to do this perfectly.

I think first order of business would be to split-out bundled libraries to be introduced each in separate commit (because +53k lines of code in one commit is rather overwhelming), preserving original author and mentioning the exact release (or commit) of upstream project, that was used. I know it might seem a bit pedantic, but this way git-blame information will be preserved, and if we'll need to introduce some changes in these files, it will be easier to resolve conflicts with upstream (if they'll ever happen).

So, my suggestion is: let's create **new pull request**, with following commits:

- 3 commit introducing only files from `dr_libs` repo, explicitly setting author info to the lead dev of that project and mentioning exact release in commit message. My suggestion would be to take the files anew, straight from `dr_libs` repo - this way if some dosbox-specific changes are needed, it will be apparent to future developers by following history of the files with git-log or gitk. Use the same file location as it is in this PR. Licensing on these libs do not require us to distribute license file in binary packages, so we don't need to worry about this for now. (these libs are ~+21k lines of code)
- 1 commit for files related to `stb_vorbis`. I have some questions regarding this library, but they likely can be trivially explained -  let's leave it for that new PR. Same as previous one, licensing does not require us to distribute document with license text. (That would be ~5k lines?)
- 1 commit for files originating from `xxHash` repo. This one is more complicated regarding to license… Upstream license in repo uses dual-licensing (BSD 2-clause or GPLv2), but the files in this PR (and same files in upstream) specify only BSD 2-clause. This means that eventually we should add a file with BSD license to the files for `make install` and `make dist` (after the lib will be hooked in the buildsystem).
- 1 commit for `archive.h`. Thanks to your comment from other file I know it comes from: https://github.com/voidah/archive, but it would be really nice to have this info in commit message.
- let's **not** include source code for SDL_sound yet - I tried correlating your changes to the history of SDL_sound sources stored in Mercurial, but couldn't find a revision that you based your changes on. I think you included some DOSBox-specific changes in there as well? SDL2_sound changed license to zlib nowadays, but it does not affect your code here.
- I think there are some other trivial changes, that could be included as well - I will point them out in initial review.

No buildsystem changes, just source of those libs for now. That review could be merged to master quite easily.

After that new PR will be merged, let's rebase this one and continue with proper review. I will leave some comments in code review now, but those are not going to be exhaustive.

File: .github/workflows/linux.yml [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    name: "Clang 8 debug (ubuntu-18.04)"
+    runs-on: ubuntu-18.04
+    steps:
+      - uses: actions/checkout@v1
+      - run: sudo ./scripts/build.sh --install-deps --compiler clang --compiler-version 8 --release debug
+      - name: "Summarize warnings"
+        run: python3 ./scripts/count-warnings.py build.log
+
+  build_ubuntu_clang_8_lto:
+    name: "Clang 8 small LTO (ubuntu-18.04)"
```
Comment (dreamer) [inline .github/workflows/linux.yml]: LTO build is cool and important for release builds, but do we need it in CI on every push?

File: .github/workflows/windows.yml [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+          cd visualc_net
+          MSBuild -m dosbox.sln -p:Configuration=Release
+
+  build_msys2_gcc_64:
+    name: MSYS2 GCC-9.2 x86_64
+    runs-on: windows-2019
+    steps:
+      - uses: actions/checkout@v1
+      - name: Install msys2 environment
+        run: choco install msys2 --no-progress
```
Comment (dreamer) [inline .github/workflows/windows.yml]: I thought msys2 was preinstalled on these machines? I need to double-check that…

File: .gitignore [hunk 3]
```diff
@@ -51,6 +51,7 @@ stamp-h1
 
 # Other
 compile
+build.log
```
Comment (dreamer) [inline .gitignore]: I think this change can be safely put into other PR and merged earlier :)

File: Makefile.am [hunk 4]
```diff
@@ -2,3 +2,25 @@
 
 EXTRA_DIST = autogen.sh
 SUBDIRS = src include docs visualc_net
+
+CLEANFILES = \
```
Comment (dreamer) [inline Makefile.am]: Maybe this can be put in a separate commit and included with other PR to be merged earlier?

File: VERSION [hunk 5]
```diff
@@ -1 +1 @@
-0.74-2
+krcroft/cdda/master
```
Comment (dreamer) [inline VERSION]: I don't know what is the purpose of this VERSION file (and value 0.74-2 indicates, that upstream does not know either - trunk branch should have value 0.75-1 here or something). Regardless, definitely it should not be changed to name of your branch name :)

File: scripts/import-from-svn/import-from-svn.sh [hunk 6]
```diff
@@ -43,7 +43,7 @@ git_svn_clone_dosbox () {
 	local -r authors_prog=$PWD/svn-authors-prog.sh
 	git -C "$repo_name" svn fetch \
 		--use-log-author \
-		--authors-prog="$authors_prog"
+		--authors-file="$authors_file"
```
Comment (dreamer) [inline scripts/import-from-svn/import-from-svn.sh]: I think this change got into your review by accident.

File: src/dos/cdrom_image.cpp [hunk 7]
```diff
@@ -16,6 +16,11 @@
  *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
  */
 
+// #define DEBUG 1
+#ifdef DEBUG
+#include <time.h>
+#include <chrono>
```
Comment (dreamer) [inline src/dos/cdrom_image.cpp]: I know rest of DOSBox code does not use correct C++-style includes, but maybe let's put `#include <ctime>` here?

`chrono` is part of C++11 :( Let's use this opportunity to ask Qbix *again* if he still does not want C++11.

File: src/dos/cdrom_image.cpp [hunk 8]
```diff
@@ -38,12 +43,32 @@
 #include <string.h>
 #endif
 
+#if defined(WORDS_BIGENDIAN)
+#define IS_BIGENDIAN true
+#else
+#define IS_BIGENDIAN false
+#endif
+
```
Comment (dreamer) [inline src/dos/cdrom_image.cpp]: I think define like this would be more appropriate in same location where `WORDS_BIGENDIAN` is defined. Or maybe it can be dropped altogether? Anyways: values `true` and `false` should be surrounded with the parentheses, as this is macro.

File: src/dos/cdrom_image.cpp [hunk 9]
```diff
@@ -38,12 +43,32 @@
...
 using namespace std;
 
 #define MAX_LINE_LENGTH 512
 #define MAX_FILENAME_LENGTH 256
 
+#ifdef DEBUG
+char* get_time() {
```
Comment (dreamer) [inline src/dos/cdrom_image.cpp]: Formatting in this function needs to be fixed.

File: src/dos/cdrom_image.cpp [hunk 10]
```diff
@@ -38,12 +43,32 @@
...
+	static time_t rawtime;
+	struct tm* ptime;
+    static char time_str[] = "00:00:00";
+
+	time(&rawtime);
+    ptime = localtime(&rawtime);
+    sprintf(time_str, "%02d:%02d:%02d", ptime->tm_hour, ptime->tm_min, ptime->tm_sec);
```
Comment (dreamer) [inline src/dos/cdrom_image.cpp]: I would use `snprintf` here. Just to be prevent an issue if someone will change formatting string, but forget to adjust `time_str`.

File: src/dos/cdrom_image.cpp [hunk 11]
```diff
@@ -70,80 +95,148 @@ int CDROM_Interface_Image::BinaryFile::getLength()
 	return length;
 }
 
-#if defined(C_SDL_SOUND)
+Bit16u CDROM_Interface_Image::BinaryFile::getEndian()
+{
+	// Image files are read into native-endian byte-order
+	#if defined(WORDS_BIGENDIAN)
```
Comment (dreamer) [inline src/dos/cdrom_image.cpp]: I *think* the style is to move all preprocessor directives to column 0 (not 100% sure though).

File: src/dos/cdrom_image.cpp [hunk 12]
```diff
@@ -313,58 +545,143 @@ bool CDROM_Interface_Image::ReadSector(Bit8u *buffer, bool raw, unsigned long se
...
+		//
+		// 3. Fill: When out circular buffer is too depleted to satisfy the
+		//	     requested size, then perform chunked-decode reads from
+		//	     the audio-codec to either fill our buffer or satify our
+		//	     remaining playback - whichever is smaller.
+		//
+		while (true) {
+
+			// 1. Consume
+			// ==========
```
Comment (dreamer) [inline src/dos/cdrom_image.cpp]: This looks trivial, but please - do remove these `======` underlines. For release builds I need to resolve small conflict in this file - when doing so, I look for git conflict markers in the file and these lines throw me off *every time*.

File: .github/workflows/linux.yml [hunk 13]
```diff
@@ -0,0 +1,51 @@
+name: "Linux builds"
+on: push
+jobs:
+
+  build_ubuntu_gcc9_debug:
+    name: "GCC 9 debug (ubuntu-18.04)"
+    runs-on: ubuntu-18.04
+    steps:
+      - uses: actions/checkout@v1
+      - run: sudo ./scripts/build.sh --install-deps --compiler-version 9 --release debug
```
[the remark this answers is not in the document]
Comment (krcroft) [reply .github/workflows/linux.yml]: I agree, but didn't have a more elegant way to bundle the two together.  One thought is to switch the action itself (--install-deps) into instructive output (--show-deps), and it would simply print the apt get install <packages> for the user to execute on their own terms.  

This would let us leverage it inside the YAML with something like:  sudo apt install $(./build.sh --show-deps), and the build script becomes more part of the code (so changes to our libs could change yet the CI YAML would stay the same)

File: .github/workflows/disabled/env.yml [hunk 14]
```diff
@@ -0,0 +1,40 @@
+name: Environment info
```
[the remark this answers is not in the document]
Comment (krcroft) [reply .github/workflows/disabled/env.yml]: Great! I threw it together to help understand what GitHub-specific (versus OS) environment variables are available or in-common across OSes.  I only have Linux at home, so this was handy to prod Windows 2016/2019 and macOS in developing the build script.

File: .github/workflows/disabled/env-macos.txt [hunk 15]
```diff
@@ -0,0 +1,760 @@
+2019-10-12T00:58:38.2405160Z Current runner version: '2.159.0'
```
[the remark this answers is not in the document]
Comment (krcroft) [reply .github/workflows/disabled/env-macos.txt]: Dgreed. Was just a reference point and indeed will go stale.  Will not include this in subsequent merges.

File: .github/workflows/linux.yml (hunk 1 above)
Comment (krcroft) [reply .github/workflows/linux.yml]: If we have a nightly build, then that's probably the best place for LTO.    

We definitely want to exercise LTO in some capacity with both compilers because it has historically been the most error-prone and difficult task for both compilers - and it's still under significantly improvement across all aspects (speed, memory, and threading). 

If code is too "assumptive", the link step of LTO will fail for various reasons.  Gentoo for years has been pushing fixes to upstream packages that fail to build with LTO, and there's been a multi-year effort by Andi Kleen to make the Linux kernel build with LTO; and it still regresses in this regard - but everyone is trying to get it there. 

The more STL we use, the more LTO will catch as well (if we're missing explicit instantiations). 
So in this regard, using LTO will catch things early when they go off the rails.

File: .github/workflows/windows.yml (hunk 2 above)
Comment (krcroft) [reply .github/workflows/windows.yml]: Only Chocolatey; there unfortunately isn't a canned msys2 VM. 
However, I think that's a good thing for us to take under control - because building a full-featured DOSBox is something that is not explicitly documented such that it's reproducible (users on vogons continue to struggle with windows builds, and are chided or told to follow the docs or "it works for me, you're probably doing something wrong").

File: src/dos/cdrom_image.cpp (hunk 8 above)
Comment (krcroft) [reply src/dos/cdrom_image.cpp]: Excellent; will make this change when we get to that "layer" in the sequence of merges.

File: src/dos/cdrom_image.cpp (hunk 11 above)
Comment (krcroft) [reply src/dos/cdrom_image.cpp]: Thanks, will make this change (and grep for anywhere else it might happen)

File: src/dos/cdrom_image.cpp (hunk 12 above)
Comment (krcroft) [reply src/dos/cdrom_image.cpp]: Will do!

**Comment by krcroft** [comment]
I like this approach; we will start from first principles and add layers of the change one merge at a time. 
Thanks for the explicit guidance on this; first commit of just the libs is ready!
We can continue to use this merge request as a discussion area to help me prep all the "layers" of this change, one merge at a time.

File: src/dos/cdrom_image.cpp (hunk 9 above)
Comment (krcroft) [reply src/dos/cdrom_image.cpp]: Thanks, will make this change.

File: src/dos/cdrom_image.cpp (hunk 10 above)
Comment (krcroft) [reply src/dos/cdrom_image.cpp]: Great suggestion, will make this change.

File: VERSION (hunk 5 above)
Comment (krcroft) [reply VERSION]: Will remove this. I was also confused if/where it's used, but I didn't want to act like I was 'releasing my own 0.74' so felt compelled to change it.

File: Makefile.am (hunk 4 above)
Comment (krcroft) [reply Makefile.am]: Yes; it's unrelated cleanup of the build system.

File: .gitignore (hunk 3 above)
Comment (krcroft) [reply .gitignore]: Yes; will bundle this in a generic PR.

File: src/dos/cdrom_image.cpp (hunk 7 above)
Comment (krcroft) [reply src/dos/cdrom_image.cpp]: Yes, there isn't much value added here. Think I'd rather use some very compelling improvements that are either far fewer lines or significantly more readable.

File: .github/workflows/disabled/env.yml (hunk 14 above)
[the remark this answers is not in the document]
Comment (krcroft) [reply .github/workflows/disabled/env.yml]: Will not add this in subsequent PRs (but will leave this conversation item open per your suggestion to pull it into the initial build step)

**Comment by dreamer** [comment]
OK, first 2 parts are merged in and I included a change from SVN, that Qbix implemented few days ago - I think it's time to rebase commit(s) in this review on top of current master branch.

File: .github/workflows/linux.yml (hunk 13 above)
[the remark this answers is not in the document]
Comment (dreamer) [reply .github/workflows/linux.yml]: I like this idea a lot - it's similar to how pkgconfig works - also as I just discovered, it could be easily reused for injecting build dependencies for rpm packages starting with Fedora 31. How about one more step and moving dependency printing into a separate script?
`sudo apt install $(./scripts/list-build-dependencies.sh)`?

File: .github/workflows/linux.yml (hunk 1 above)
Comment (krcroft) [reply .github/workflows/linux.yml]: Knocked out the debug and LTO builds, in favor of using those in the nightly build & asset packager.

File: .github/workflows/linux.yml (hunk 13 above)
[the remark this answers is not in the document]
Comment (krcroft) [reply .github/workflows/linux.yml]: Nice and tidy; will do!
The expansion approach works great under brew and apt, but I can't the nested shell expansion to work properly in Windows. 

Here's what I've tried so far (in the YAML):

`run: C:\tools\msys64\usr\bin\bash -lc "pacman -S --noconfirm $($GITHUB_WORKSPACE/scripts/build.sh --list-packages --bin-path /mingw64/bin)"`

That threw syntax errors on the first $, so I decided to switch to the (not-recommended) back-tick sub-shell:

```
run: C:\tools\msys64\usr\bin\bash -lc "pacman -S --noconfirm `$GITHUB_WORKSPACE/scripts/build.sh --list-packages --bin-path /mingw64/bin`"
```

But that also failed.

I gave up so far and have explicitly listed the packages.  Need to figure something out.

File: src/dos/cdrom_image.cpp (hunk 8 above)
Comment (krcroft) [reply src/dos/cdrom_image.cpp]: Figured out a way around using macros; it's pretty clean. You'll see it when we get to the layered PRs.  (I'm still working on this branch as my main sandbox to keep testing everything)

File: src/dos/cdrom_image.cpp (hunk 11 above)
Comment (krcroft) [reply src/dos/cdrom_image.cpp]: They're all moved now.

**Comment by krcroft** [comment]
> OK, first 2 parts are merged in and I included a change from SVN, that Qbix implemented few days ago - I think it's time to rebase commit(s) in this review on top of current master branch.

Right on - let me know what you're thinking for the next PR and the commits!

File: .github/workflows/linux.yml (hunk 13 above)
[the remark this answers is not in the document]
Comment (dreamer) [reply .github/workflows/linux.yml]: Windows uses cmd.exe as default shell, it can be changed to PowerShell or bash.

[the remark this answers is not in the document]
Comment (dreamer) [reply .github/workflows/linux.yml]: Example:

    steps:
      - name: Display the path
        run: echo $PATH
        shell: bash

Link to documentation: https://help.github.com/en/github/automating-your-workflow-with-github-actions/workflow-syntax-for-github-actions#jobsjob_idstepsrun

**Comment by dreamer** [comment]
At +7.6k lines it's much easier to review code already - previously GitHub interface was a bit slow :)

I think the next part, that could be easily merged would probably be build scripts + build.md and CI changes (I think these changes are big enough to put them in separate commits).

After CI, another part could be a dump of original files from SDL_audio (before your changes, with authorship set to Ryan C. Gordon). I think these will be only 3-5 files at this point? (SDL_sound.h, SDL_sound.c, SDL_sound_internal.h). There's also file audio_convert.c - but I'm not sure if it was part of SDL or did you include it later? I will try again to find out what was the original revision that you based your changes on.

**Comment by dreamer** [comment]
Update to my previous comment from 7h ago (since GitHub so helpfully decided to hide it):
I found exact version of SDL_sound you based your changes on: https://github.com/dreamer/dosbox-staging/pull/20

**Comment by krcroft** [comment]
A couple more PRs ready for review!
 - CI/build tools
 - SDL_Sound updates

The compiler-warning cleanup changes might be of interest to upstream. Back in 2017 I sent details of adding Opus to the SDL mailing list, and can do the same for this.

**Comment by krcroft** [comment]
@dreamer , not many files left in the CD Audio set! 
For the next PR, I could add the SDL decoders sources (opus.c, vorbis.c, mp3.cpp and the fast seek table, ...). 
That would leave the last PR with the few changes tying it into to the existing DOSBox sources like cdrom_image and the header, configure.ac and the two Makefile.am files.

**Comment by dreamer** [comment]
> @dreamer , not many files left in the CD Audio set!
> For the next PR, I could add the SDL decoders sources (opus.c, vorbis.c, mp3.cpp and the fast seek table, ...).
> That would leave the last PR with the few changes tying it into to the existing DOSBox sources like cdrom_image and the header, configure.ac and the two Makefile.am files.

Sounds like a plan :)

I am planning to send 2nd set of small patches to upstream (hopefully tonight) - do you want me to include changes, that were merged to `src/libs`, or do you prefer to send them yourself?

**Comment by krcroft** [comment]
I like the idea of you sending patches from master, just from a consistency standpoint (ie: Generated the same way, you will include them all as attachments in one message, and Qbix and others can step through one by one). 

Also, if you think this layered approach to reviewIng and accepting also makes sense (I believe it will help upstream). I can see how the entire ball of cdda work would be overwhelming and your PR approach makes a lot of sense. 

So yes, would definitely appreciate this!

**Comment by dreamer** [comment]
I wonder how many lines are left in this review after rebase to the newest master :)

**Comment by krcroft** [comment]
> I wonder how many lines are left in this review after rebase to the newest master :)
Not many! We're about to find out :-)

**Comment by dreamer** [comment]
@krcroft I just submitted non-CI related commits as a zipfile with patches to sourceforge: https://sourceforge.net/p/dosbox/patches/284/ - please take part in discussion, once it'll start :)

**Comment by krcroft** [comment]
Thank you for this. Your submission looks great, and I'm looking forward to the discussion.

**Comment by dreamer** [comment]
Part 4 is merged… I think the bulk part of code, that constituted this review is in? Perhaps it's time to close this PR?

Part 5 covers #27, right?

**Comment by krcroft** [comment]
> Part 4 is merged… I think the bulk part of code, that constituted this review is in? Perhaps it's time to close this PR?
> 
> Part 5 covers #27, right?

Yes to both; Part 4 completes the work and covers the patch as it's been (functionally) for quite some time (tested heavily under ECE). It also stays within the confines of how SDL_Sound, the Mixer, and CDROM image "traditionally" operate. 

With part 5, all three aspects are now being reshaped to suite our own needs and cut out lots of intermediate code/buffers/behavior.
